apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/Numeric.cs

```diff
@@ -0,0 +1,663 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Collections.Concurrent;
+using System.Globalization;
+using System.Linq;
+using System.Numerics;
+using System.Text.RegularExpressions;
+
+namespace Google.Cloud.BigQuery.V2
+{
+    // Note: These types are internal and used for implementing BigQueryNumeric and
+    // BigQueryBigNumeric only, which are thoroughly tested.
+    // If we ever need to make these types public, then we need to test them on its own.
+    internal sealed class NumericType
+    {
+        private sealed class Parameters
+        {
+            internal Parameters(BigInteger scaledMinValue, BigInteger scaledMaxValue, int scale)
+            {
+                GaxPreconditions.CheckArgumentRange(scale, nameof(scale), 0, PowersOf10.Length - 1);
+                GaxPreconditions.CheckState(scaledMinValue <= scaledMaxValue,
+                    $"{nameof(scaledMinValue)}={scaledMinValue} is greater than {nameof(scaledMaxValue)}={scaledMaxValue}.");
+
+                Scale = scale;
+                ScaledMinValue = scaledMinValue;
+                ScaledMaxValue = scaledMaxValue;
+            }
+
+            // These are scaled by the scaling factor.
+            // A value of 1 is really 0.000000001 if the scale is 9.
+            public BigInteger ScaledMinValue { get; }
+            public BigInteger ScaledMaxValue { get; }
+
+            public int Scale { get; }
+
+            public override bool Equals(object obj) =>
+                obj is Parameters def
+                && def.ScaledMinValue.Equals(ScaledMinValue)
+                && def.ScaledMaxValue.Equals(ScaledMaxValue)
+                && def.Scale.Equals(Scale);
+
+            public override int GetHashCode() =>
+                19 * Scale.GetHashCode()
+                + 59 * ScaledMinValue.GetHashCode()
+                + 79 * ScaledMaxValue.GetHashCode();
+        }
+
+        internal const int MaxSupportedScale = 38;
+
+        // This is a helper value. The length of the array depends on the supported precisions and scales.
+        // The current length (39) is enough to support BigQueryNumeric and BigQueryBigNumeric.
+        internal static readonly BigInteger[] PowersOf10 = Enumerable.Range(0, 39).Select(p => BigInteger.Pow(10, p)).ToArray();
+
+        // TODO: Don't require a 0 before the decimal point.
+        // TODO: Replace with manual validation if we find this is a performance bottleneck (as it was with field name validation).
+        private static readonly Regex s_validation = new Regex(@"^-?[0-9]+\.?[0-9]*$");
+
+        private static readonly ConcurrentDictionary<Parameters, NumericType> s_defined = new ConcurrentDictionary<Parameters, NumericType>();
```

Comment:
Do we actually need a cache here? If we don't make this public, we'll only have a set of a few instances of `NumericType`... and they could be created in the relevant types.

Let's chat about this.